oslo_vmware/rw_handles.py

```diff
@@ -658,6 +738,7 @@ def close(self):
         :raises: VimException, VimFaultException, VimAttributeException,
                  VimSessionOverLoadException, VimConnectionException
         """
+        self._conn.close()
```

Comment:
This is already achieved down in the `finally`. If you follow the path it leads to `FileHandle.close()` which closes the same socket (`self.file_handle`).

Reply:
Actually, no. That is the rather subtle bug this change is trying to fix. I think, the confusion stems mostly from the wrong naming of functions and variables.

The short of it:
* `FileHandle._file_handle` was and is still referring to an `HTTPResponse` (not a socket/HTTPConnection) for the reading handles (in contrast to the writing ones)
* Prior to this change:
  *    _create_read_connection was returning the *`HTTPResponse`* (despite having connection in its name) and
  *   `VmdkReadHandle._conn` was also pointing to a *`HTTPResponse`* (again, despite its name, not a HTTPConnection), so your point would have been correct
* After the change, `_conn` is pointing to the `HTTPConnection` from which the response comes from

Let me run you through it in more detail:
Since we need access to the socket to close it, I changed  [_create_read_connection](https://github.com/sapcc/oslo.vmware/pull/23/files#diff-6240dfe7c5adee5473b15d76a6ebc93611bad8ad879146b42d27b1ff9b7ee1b0L103-R106) to behave the same way as its sibling `_create_write_connection` and its name would suggest, and return an `HTTPConnection`. 

As you can see in the diff, originally the `HTTPConnection` was never returned, much less stored. 
Instead the `HTTPResponse` was stored in the confusingly named member `self._conn` and passed on to `FileHandle`, where it would then be stored as `FileHandle._file_handle`. That response object would then be closed through the code path you pointed out. 

After the change, `_conn` is a `HTTPConnection`, but since the existing code is presumably relying on the `_file_handle` being a response, I now get the response from the connection and pass it to `FileHandle` [here](https://github.com/sapcc/oslo.vmware/pull/23/files#diff-6240dfe7c5adee5473b15d76a6ebc93611bad8ad879146b42d27b1ff9b7ee1b0R709), to be stored as `FileHandle.file_handle` as before.

Now, `self._conn` refers to the connection, and not anymore to the response, while `self._file_handle` still refers to the response.
The line you are pointing out is now closing the connection, which would otherwise be closed at an arbitrary later time,  while `FileHandle.close()` still closes the response-object.

I have a simple test-case which demonstrates the effect:
- Detach a CD-ROM
- Download the ISO via the API
- Reattach the CD-ROM right after

Prior this change, it was raising a locking error and wasn't possible.
Looking at the code, I was seeing that only the response was closed and the connection untouched.
This led me suspect that the API from the ESXi is keeping a lock on the file when keeping an connection open, even if the request is finished.
After this change, I never did run into the locking error again.

Reply:
Thanks for explaining it. Looking at it again it makes total sense.

Looking at python docs, `HTTPResponse` doesn't even have a `.close()` method (at least not before 3.5). Does it still make sense then to call the `super()` for closing the `_file_handle` ?

Reply:
Apparently the python docs are [incomplete]( https://github.com/python/cpython/blob/2.7/Lib/httplib.py#L557-L561).
I was looking it up, because I was otherwise expecting an exception, which I haven't seen.

I think, it still makes sense to close `super().close()`, the parent still might need to do some cleanup, and closing the response seems to clean up some buffer, and maybe even is necessary to close the socket properly. I am not clear about that.